docs/clasp.1
.\"
.\" man page introduction to Clasp
.\"
.\" Copyright (c) 2024, Clasp Developers
.\"           All Rights Reserved
.TH CLASP 1 "20 June 2024"
.SH NAME
Clasp -- A Common Lisp native compiler using LLVM
.SH SYNOPSIS
\fBclasp\fP [ --help ] [ --noinform ] [ --noprint ] [ --disable-debugger ]
[ --non-interactive ] [ --feature \~\fIfeature\fP ]
[ --eval \fIform\fP | --load \~\fIfilename\fP ]*
[ --norc ] [ --version ] [ \~\fIother options\fP ]

.SH DESCRIPTION

\fBClasp\fP is an implementation of Common Lisp primarily designed for
compatibility with C++-language programs and libraries.

Supported operating systems are Linux distributions Arch, Debian/Ubuntu, Fedora,
Gentoo, and also MacOS. Its requirements (build and runtime) are primarily LLVM
(versions 15 through 18), and several of the standard Common Lisp libraries.

.SH RUNNING CLASP

Running clasp without any arguments starts the interactive Read / Evaluation /
Print Loop (REPL). At the prompt you may enter a Lisp expression; it will be
executed, any values it produces will be printed, and you will recieve another
prompt allowing you to enter another expression.

\fBTo exit clasp\fP enter the expression "\f(CR(ext:quit [integer])\fR" (and
return), or type a ^D (control-D).\fP The integer argument, if provided, is used
as the exit status; the default is zero. \fI(N.B. The shell will treat 0 as
true, non-zero as false/error.)

\f(CR
  $ clasp
  ...[startup banner]...
  COMMON-LISP-USER> (* 6 7)

  42
  COMMON-LISP-USER> (ext:quit)
\fR

.SH COMMAND LINE OPTIONS

This is a summary of the most common options. The \f(CR\-\-help\fR option
will display the full list of supported options, environment variables,
and \f(CR*features*\fR flags.
.TP 3
.B \-\-noinform
Skip messages at startup.
.TP 3
.B \-\-noprint
Start a read-eval loop instead of a read-eval-print loop, and don't prompt. This
is intended for scripts (\fIe.g.\fP piping a list of forms to Clasp).
.TP 3
.B \-\-disable-debugger
Set things up so that if the default debugger would be entered, Clasp quits with
a backtrace and nonzero exit status instead. \f(CR*debugger-hook*\fR and
\f(CRext:*invoke-debugger-hook*\fR (below) work as usual, \fIi.e.\fR this
setting does not affect them.
.TP 3
.B \-\-non-interactive
Rather than starting a REPL, quit. This is intended to be used with \f(CR\-\-eval\fR and \f(CR\-\-load\fR. Implies \f(CR\-\-disable-debugger\fR.
.TP 3
.B \-\-feature \~\fIfeature\fP
Intern "\fIfeature\fP" as a keyword and push it to \f(CR*features*\fR.
.TP 3
.B \-\-eval \~\fIform\fP
Evaluate the given form. \f(CR\-\-eval\fR and \f(CR\-\-load\fR options are
processed in order from left to right.
.TP 3
.B \-\-load \~\fIfilename\fP
\f(CRcl:load\fR the given file. Intended for convenience, since \f(CR--eval
(load ...))\fR could involve annoying quotation issues.
.TP 3
.B \-\-norc
Do not load \f(CR~/.clasprc\fR (see below).
.TP 3
.B \-\-help
Displays a full list of options, environment variables, and feature flags.
.TP 3
.B \-\-version
Displays the program version string and exits.
.PP
If the file \f(CR~/.clasprc\fR exists, Clasp will \f(CRcl:load\fR it before
processing \f(CR\-\-eval\fR and \f(CR\-\-load\fR options and starting the REPL.

.SH CREDITS

Clasp is the project of Dr. Christian Schafmeister. Additional contributions
have been made by Alex Wood, Karsten Poeck, and many other contributors as seen
in the Git history.

Clasp's source code is derived substantially from that of Embeddable Common
Lisp. Code from SBCL and SICL has been incorporated as well. Most notably, the
compiler is SICL's Cleavir compiler with some minor customizations.

.SH ENVIRONMENT

See the \f(CR\-\-help\fR option for a complete list.

.TP 10n
.BR CLASP_DEBUG \~\fIfilenames\fP
Define files that generate log info when \f(CRDEBUG_LEVEL_FULL\fR is set at
top of file. \~\fIfilenames\fP is separated with spaces or commas.
.TP 10n
.BR CLASP_HOME \~\fIdir\fP
Define where clasp source code lives.
.TP 10n
.BR CLASP_FEATURES \~\fIfeatures\fP
Set \f(CR*features*\fR (separate multiple features with spaces or commas).

.SH FILES

.TP
.I ~/.clasprc
Optional per-user start-up script.

.SH SEE ALSO

Clasp project home page
.IP
<\f(CRhttps://github.com/clasp-developers/clasp/tree/main\fR>
.PP
Clasp manual
.IP
<\f(CRhttps://github.com/clasp-developers/clasp/wiki/Manual\fR>
.PP
Building from source
.IP
<\f(CRhttps://github.com/clasp-developers/clasp/wiki/Building-and-Installing-from-Source\fR>
.PP
Real-time chat
.IP
\f(CR#clasp\fR IRC channel on Libera <\f(CRhttps://web.libera.chat\fR>
.PP
.SH LICENSE

CLASP is free software; you can redistribute it and/or modify it
under the terms of the GNU Library General Public License as published
by the Free Software Foundation; either version 2.1 of the License, or
(at your option) any later version; see file 'Copying'.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Library General Public License for more details.

You should have received a copy of the GNU Library General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


RELEASE-NOTES
Upcoming Clasp version 1.0.0

* New compiled library format called FASP - it uses concatenated object files.
* Implemented save-lisp-and-die. This saves the state of a running environment for loading and fast startup later. 
Our most complex environment Cando starts up in ~4 seconds, which is 10x faster than the old startup that loaded libraries.

Clasp version 0.5.0

* Full integration of the Cleavir/Sicl compiler by Robert Strandh (https://github.com/robert-strandh/SICL)
* A nascent type inference system within Cleavir by Alex Wood
* LLVM whole program, Link-Time-Optimization
* Multi threading compatible with Bordeaux threads
* Unicode, 32bit wide characters
* Specialized arrays for many primitive types (int8, int16, int32, int64, double, float)
* CFFI (C Foreign Function Interface)
* New build system based on waf (https://github.com/waf-project/waf)
* Code and file cleanup
* Better Emacs and Slime support including jumping to source (C++ and CL),
  tab-completion and lambda-list display
* Many, many improvements and bug fixes


Clasp version 0.4.0

* Clasp has a completely new compiler!
  The Cleavir Common Lisp compiler developed by Robert Strandh as part of his SICL project.
  Cleavir is an advanced compiler that generates fast code
  and is an excellent compiler for testing advanced optimization algorithms.
  Cleavir carries out escape analysis to properly assign variables to the stack/heap
  and it does aggressive inlining. The best code that Cleavir generates has clocked within
  a factor of four of the speed of C.
  For more details on SICL see: https://github.com/robert-strandh/SICL
* Fixnum, character and single-float types are now immediate values and fixnums are tagged
  using #b000 in the bottom three bits to allow fast integer arithmetic.
* General object pointers are tagged, cons cells have a separate tag to accelerate are tagged.
* Building Clasp has been accelerated and now takes about 2.5 hours on a reasonably powerful
  Linux system and 3.5 hours on OS X.
* Clbind library allows programmers to expose external C++ libraries to Clasp Common Lisp.
* The build system has been improved and externals-clasp has been eliminated as a requirement.
* Clasp version 0.4.0 is a Boehm only version of Clasp.
  The Memory Pool System garbage collector
  is being tuned for Clasp and is planned to be in the 0.5.0 release.



Clasp version 0.11

* Added ASDF support.  
  This is still alpha.  Compile the ASDF module using (core:compile-asdf).
  After that you can load the module using (load "sys:kernel;asdf;build;asdf.bundle").
  It takes between 15-30 seconds to load (this is why I'm integrating Cleavir).
* Added the :CLASP *feature* and removed the :ECL *feature*.
  Clasp will continue to mimic the underlying ECL functionality so that
  Common Lisp code that supports ECL can be made to support Clasp by converting
  #+ecl to #+(or ecl clasp)   and #-ecl to #-(or ecl clasp)
* Added code to generate object files directly from Clasp.
  The LLVM bitcode compiler "llc" no longer needs to be in the PATH
  for Clasp to generate object files from Common Lisp source.
  The "ld" linker does need to be accessible.



guix/clasp.scm
;;; SPDX-FileCopyrightText: Copyright (c) 2024 The Clasp Authors
;;; SPDX-License-Identifier: MIT
;;;
;;; (Of course, Guix is GPL-3.0-or-later.)

;; Usage:
;;
;;   guix shell -D -f guix/clasp.scm -- COMMAND # Clasp-suitable environment
;;   guix build -f guix/clasp.scm               # Build the package
;;   guix package -f guix/clasp.scm             # Install the package
;;
;; Unfortunately, for now the package and the above commands require the
;; cloned dependencies to be already present, which can be accomplished
;; as follows:
;;
;;   guix shell --pure git nss-certs sbcl -- ./koga
;;
;; though Koga will error out after downloading the dependencies, when
;; trying to configure Clasp.

(use-modules
 (ice-9 ftw)
 (srfi srfi-1)
 (srfi srfi-11)
 (git bindings)
 (git describe)
 (guix build-system gnu)
 (guix gexp)
 (guix git-download)
 ((guix licenses) #:prefix license:)
 (guix packages)
 (gnu packages base)
 (gnu packages boost)
 (gnu packages certs)
 (gnu packages compression)
 (gnu packages elf)
 (gnu packages gcc)
 (gnu packages libunwind)
 (gnu packages lisp)
 (gnu packages llvm)
 (gnu packages multiprecision)
 (gnu packages ninja)
 (gnu packages pkg-config)
 (gnu packages pretty-print)
 (gnu packages version-control))

(define (git-checkout-description checkout)
  (libgit2-init!)
  (let-values (((commit description) (describe-checkout checkout)))
    description))

(define (scandir-absolute name)
  (let ((contents (scandir name (lambda (file)
                                  (not (or (equal? "." file)
                                           (equal? ".." file)))))))
    (if (not contents) '()
        (map-in-order (lambda (file) (string-append name "/" file))
                      contents))))

(define clasp-cl
  (let* ((source-dir (dirname (dirname (current-filename))))
         (version (git-checkout-description source-dir))
         (external-dirs               ; dirs containing nested Git repos
          (list (string-append source-dir "/src/lisp/kernel/contrib")
                (string-append source-dir "/dependencies")
                (string-append source-dir "/extensions")))
         (git-dirs
          (apply append
                 (list (string-append source-dir "/src/bdwgc")
                       (string-append source-dir "/src/libatomic_ops")
                       (string-append source-dir "/src/lisp/modules/asdf"))
                 (map-in-order scandir-absolute external-dirs)))
         (descend-dirs (append git-dirs external-dirs))
         (predicates (map-in-order git-predicate (cons source-dir git-dirs))))
    (package
      (name "clasp-cl")
      (version version)
      (source
       (local-file source-dir (git-file-name "clasp-cl" version)
                   #:recursive? #t
                   #:select?
                   (lambda (file stat)
                     (or (any (lambda (dir) (string=? dir file)) descend-dirs)
                         (any (lambda (pred) (pred file stat)) predicates)))))
      (build-system gnu-build-system)
      (arguments
       (list
        #:configure-flags
        #~(let ((clang #$(this-package-input "clang-toolchain")))
            (list "--build-mode=bytecode-faso"
                  "--reproducible-build"
                  (string-append "--bin-path=" #$output "/bin")
                  (string-append "--lib-path=" #$output "/lib/clasp")
                  (string-append "--share-path=" #$output "/share/clasp")
                  ;; Without --cc and --cxx, Clang is searched for in
                  ;; LLVM's /bin.
                  (string-append "--cc=" clang "/bin/clang")
                  (string-append "--cxx=" clang "/bin/clang++")
                  (string-append
                   "--ld=" #$(this-package-native-input "lld") "/bin/ld.lld")
                  (string-append
                   "--ldflags="
                   "-Wl,-rpath," clang "/lib"
                   " -Wl,-rpath," (ungexp (this-package-input "gcc") "lib") "/lib"
                   " -Wl,-rpath," #$(this-package-input "fmt") "/lib"
                   " -Wl,-rpath," #$(this-package-input "gmp") "/lib"
                   " -Wl,-rpath," #$(this-package-input "libelf") "/lib")))
        #:phases
        #~(modify-phases %standard-phases
            (add-after 'unpack 'fix-paths
              (lambda _
                (substitute* '("dependencies/quicklisp-client/asdf.lisp"
                               "src/lisp/modules/asdf/uiop/run-program.lisp"
                               "src/lisp/modules/asdf/uiop/launch-program.lisp"
                               "src/lisp/regression-tests/extensions.lisp")
                  (("\"/bin/sh\"")
                   (string-append "\"" (which "sh") "\"")))))
            (delete 'bootstrap)
            (replace 'configure
              (lambda* (#:key (configure-flags '()) #:allow-other-keys)
                (setenv "HOME" "/tmp")
                (apply invoke "sbcl" "--script" "koga" "--skip-sync"
                       "--jobs=1"       ; overridden in build phase
                       configure-flags)))
            (replace 'build
              (lambda* (#:key parallel-build? #:allow-other-keys)
                (let ((jobs (if parallel-build?
                                (number->string (parallel-job-count))
                                "1")))
                  (setenv "CLASP_BUILD_JOBS" jobs)
                  (invoke "ninja" "-C" "build" "-j" jobs))))
            (replace 'check
              (lambda* (#:key tests? #:allow-other-keys)
                (when tests?            ; always serial
                  (invoke "ninja" "-C" "build" "test"))))
            (replace 'install
              (lambda _
                (invoke "ninja" "-C" "build" "install"))))))
      (native-inputs
       (list binutils
             coreutils
             git
             lld
             ninja
             pkg-config
             sbcl
             ;; for koga --archive
             gzip nss-certs tar))
      (inputs
       (list boost
             clang-toolchain-18
             fmt
             (list gcc "lib")
             gmp
             libelf
             libunwind
             libunwind-headers))
      (home-page "https://clasp-developers.github.io/")
      (synopsis "Common Lisp implementation based on LLVM")
      (description "Clasp is a Common Lisp implementation based on LLVM that is
designed to interoperate with C++ libraries.")
      ;; Various files are licensed under permissive licenses.
      (license license:lgpl2.1+))))

clasp-cl
